lib/crypto/c_src/algorithms.c

```diff
@@ -190,43 +195,79 @@ ERL_NIF_TERM mac_algorithms(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 
 ERL_NIF_TERM curve_algorithms(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 {
+    int fips_mode = 0;
+    int algo_curve_cnt = 0;
+
 #ifdef FIPS_SUPPORT
-    if (FIPS_mode())
-        return enif_make_list_from_array(env, algo_curve[1], algo_curve_fips_cnt);
-    else
+    fips_mode = FIPS_mode();
 #endif
-    return enif_make_list_from_array(env, algo_curve[0], algo_curve_cnt);
+    algo_curve_cnt = get_curve_cnt(env, fips_mode);
+    return enif_make_list_from_array(env, algo_curve[fips_mode], algo_curve_cnt);
 }
 
 #if defined(HAVE_EC)
 int init_curves(ErlNifEnv* env, int fips);
 int valid_curve(int nid);
 #endif
 
+int get_curve_cnt(ErlNifEnv* env, int fips) {
+    static unsigned int algo_curve_cnt, algo_curve_cnt_initialized;
+    static unsigned int algo_curve_fips_cnt, algo_curve_fips_cnt_initialized;
+    int cnt = 0;
+    if (0 == fips  && 1 == algo_curve_cnt_initialized) {
+        return algo_curve_cnt;
+    }
+
+    if (1 == fips  && 1 == algo_curve_fips_cnt_initialized) {
+        return algo_curve_fips_cnt;
+    }
+
+    enif_mutex_lock(mtx_init_curve_types);
+    if (1 == fips) {
+        if (1 == algo_curve_fips_cnt_initialized) {
+            return algo_curve_fips_cnt;
+        }
+
+#if defined(HAVE_EC)
+        algo_curve_fips_cnt = init_curves(env, 1);
+#endif /* defined(HAVE_EC) */
+        cnt = algo_curve_fips_cnt;
+        algo_curve_fips_cnt_initialized = 1;
+    } else {
+        if (1 == algo_curve_cnt_initialized) {
+            return algo_curve_cnt;
+        }
+
+#if defined(HAVE_EC)
+        algo_curve_cnt = init_curves(env, 0);
+#endif /* defined(HAVE_EC) */
+        cnt = algo_curve_cnt ;
+        algo_curve_cnt_initialized = 1;
+    }
+    enif_mutex_unlock(mtx_init_curve_types);
+
+    return cnt;
+}
+
 void init_curve_types(ErlNifEnv* env) {
-    algo_curve_cnt = algo_curve_fips_cnt = 0;
+    int curve_cnt = 0;
```

Comment:
Here we get the warning:

variable ‘curve_cnt’ set but not used [-Wunused-but-set-variable] int curve_cnt = 0;